samples/performance/constant_data/constant_data_tutorial.md

```diff
@@ -0,0 +1,405 @@
+
+***Note:** If deciding to use a `uniform`, then the size of the array needs to be defined at compile time. This can be achieved with a shader variant definition.*
+
+Here is an example of using a `buffer` in shader code:
+
+```
+layout(set = 0, binding = 1) buffer MeshArray
+{
+	mat4 model_matrices[];
+} mesh_array;
+```
+
+***Note:** If you plan to write to `buffer`s inside shaders, you have to query the physical device features and check that `vertexPipelineStoresAndAtomics` is enabled.*
+
+Before you draw the scene, you create a `VkBuffer` with the  `VK_BUFFER_USAGE_STORAGE_BUFFER_BIT` usage flag, and fill it with all the model matrices of each mesh in the scene.
+
+Then to get the correct matrix inside our shader, we can pass a **dynamic index** to our draw call. We do this by using the `gl_InstanceIndex` value.
+
+For example, your shader code will look something like this:
+
+```
+mat4 model_matrix = mesh_array.model_matrices[gl_InstanceIndex];
+
+out_pos = model_matrix * vec4(in_pos, 1.0);
+```
+
+To control the value of `gl_InstanceIndex` we use the `uint32_t firstInstance` parameter of the `vkCmdDraw*()` commands.
+
+It's important to note that we can use other mechanisms to push this index to the shader, such as push constants.
+
+For example, this `vkCmdDrawIndexed` is taken from the [Vulkan spec](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkCmdDrawIndexed.html):
+
+```
+void vkCmdDrawIndexed(
+    VkCommandBuffer                             commandBuffer,
+    uint32_t                                    indexCount,
+    uint32_t                                    instanceCount,
+    uint32_t                                    firstIndex,
+    int32_t                                     vertexOffset,
+    uint32_t                                    firstInstance);
+```
+
+Here is some pseudo-code to show how the `vkCmdDrawIndexed` function is used, and also to describe how a generic scene render function will look:
+
+```
+uint32_t instance_index = 0;
+
+vkCmdBindDescriptorSet(command_buffer, buffer_array_descriptor_set);
+
+for(auto &mesh : meshes)
+{
+    vkCmdBindVertexBuffer(command_buffer, mesh.vertex_buffer);
+    vkCmdBindIndexBuffer(command_buffer, mesh.index_buffer);
+
+    // This line does our drawing
+    vkCmdDrawIndexed(command_buffer, mesh.index_buffer.size(), 1, 0, 0, instance_index++);
+}
+```
+
+In the code snippet above we can see that we bind our descriptor set once, and for each mesh bind its vertex and index buffers and then execute a draw call with an incrementing value for `uint32_t firstInstance`. This `uint32_t` will be substituted in wherever `gl_InstanceIndex` exists in the shader code, which will pull out the required model matrix to position the mesh inside our world.
+
+
+### **Performance**
+
+This on Mali is not a recommend practice as it disables a compiler optimisation technique known as **pilot shaders**. 
```

Comment:
This is one of the fastest methods for us, similar to push-constant performance.